### ISSUE TITLE
remove gogo/protobuf direct dependency

### ISSUE THREAD

**Issue by rboyer** [hashicorp/consul#10471]
The `github.com/gogo/protobuf` library is [now unmaintained](https://github.com/gogo/protobuf/issues/691) and no longer tracks with the latest major release of `github.com/golang/protobuf` (the 1.4.x series). The two will _sorta_ interoperate up to a point, and that point manifests as marshal/unmarshal `panic` which are very undesirable.

All of the protobufs and generated code under the `./proto` directory are generated with the extended options offered by gogo, with features like optionally embedding structs, using `time.Time` or `time.Duration` in generated code, and adding struct tags to fields.

Unfortunately as gogo/protobuf is effectively dead, we have to cease using all of the bells and whistles it provided before we can laterally go back to direct use of `golang/protobuf` (1.3.x) and after that migration can bump to the 1.4.x series (or jump to the new repo where v2 is maintained).

Some of these steps can happen in incremental commits, but unfortunately due to the cross-cutting nature of these changes many of these will have to be monolithic singular commits.

1. Merge improvements (https://github.com/hashicorp/consul/pull/10229) to the `mog` tool and move it to its own repo to make the tool easier to use, rather than having it hide in an unmerged branch of this repo.
2. Selectively go through various `./proto` files and remove gogo extensions, regenerate the code, and fix anything that is broken by those removals.
    1. Likely any place we use the generated protobuf structs and expect to marshal to/from json into those will need to stop, since we rely upon stuff like embedded structs (like the raft fields) and struct tags (for `bexpr` filtering or `hashstructure` stuff) to get the "shape" of the json to be what our API package expects. For this we can just use `mog` to transmogrify the protobuf structs into either `api` package structs or new hand-rolled structs in `agent/structs` as is appropriate.
    2. Fix how `time.Time` and `time.Duration` work. Noting that the protobuf time structs have less precision than the stdlib structs so various `auto-config` or `agent/connect` CA tests will fail due to this and need precision truncation to correct.
3. Once we aren't using gogo extensions for any protobufs, laterally moving over to the version of `golang/protobuf` should be straightforward.
    1. We **likely** do not want to upgrade the pinned version of `protoc` yet in this procedure until we're ready for 1.4.x.
    2. ...unless gRPC has compilation errors somewhere due to some back compat issues they had in which case try bumping `protoc` by very tiny  semver version increments until it isn't broken.
4. After there's no more gogo you can take a break and start tackling direct official protobuf version upgrades in another issue.

**Comment by rboyer** [comment]
There's an incomplete (and likely now too old to merge safely) reference PR I can privately share to intrepid explorers.

**Comment by dnephin** [comment]
This issue (removing gogo/protobuf) is blocking us from upgrading gRPC. More recent versions of gRPC require a version of protobuf beyond what gogo/protobuf supports (if my understanding is correct). Leave a note here to remind myself the next time this comes up.

**Comment by jkirschner-hashicorp** [comment]
@rboyer : which PR closed this?

**Comment by rboyer** [comment]
https://github.com/hashicorp/consul/pull/12639
